ui/app/js/controllers/required_signoffs_modal_base_controller.js

```diff
@@ -0,0 +1,349 @@
+angular.module("app").controller("BaseRequiredSignoffCtrl",
+function($scope, $modalInstance, $q, CSRF, ProductRequiredSignoffs, PermissionsRequiredSignoffs, Rules,
+         required_signoffs, mode, product, channel, current_user, current_roles, editing) {
+  $scope.saving = false;
+  $scope.errors = {};
+
+  $scope.mode = mode;
+  $scope.product = product;
+  $scope.channel = channel;
+  // When saveChanges is called, new_roles contains whatever the user has entered.
+  // We need the original version as well, so we can compare the two and figure
+  // out how to move from the current state to the new state.
+  $scope.new_roles = $.extend(true, [], current_roles);
+  $scope.current_roles = current_roles;
+  $scope.editing = editing;
+
+  $scope.products = [];
+  Rules.getProducts().success(function(response) {
+    $scope.products = response.product;
+  });
+  $scope.channels = [];
+  Rules.getChannels().success(function(response) {
+    $scope.channels = response.channel;
+  });
+
+  $scope.length = function(item) {
+    return item.length;
+  };
+
+  $scope.getTitle = function () {
+    var title = "Signoff Requirements";
+    if ($scope.product !== "") {
+      if ($scope.mode === "channel" && $scope.channel !== "") {
+        title += " for the " + $scope.product + " " + $scope.channel + " channel";
+      }
+      else if ($scope.mode === "permissions") {
+        title += " for " + $scope.product + " Permissions";
+      }
+    }
+    return title;
+  };
+
+  $scope.addRole = function() {
+    $scope.new_roles.push({"role": "", "signoffs_required": null, "sc": null, "new": true});
+  };
+
+  $scope.removeRole = function(index) {
+    $scope.new_roles.splice(index, 1);
+  };
+
+  $scope.saveChanges = function() {
+    $scope.errors = {};
+    var service = null;
+
+    if (Object.keys($scope.new_roles).length === 0) {
+      $scope.errors["exception"] = "No roles found!";
+      return;
+    }
+
+    if ($scope.mode === "channel") {
+      service = ProductRequiredSignoffs;
+    }
+    else if ($scope.mode === "permissions") {
+      service = PermissionsRequiredSignoffs;
+    }
+    else {
+      $scope.errors["exception"] = "Couldn't detect mode";
+      return;
+    }
+
+    $scope.saving = true;
+    CSRF.getToken()
+    .then(function(csrf_token) {
```

Comment:
Thanks for all the help refactoring this. I _think_ it's a bit better now, but I'm curious what you think.